sal/tests/client/functional/test_operations.py

```diff
@@ -0,0 +1,243 @@
+"""
+Tests the SAL operations: list, get, put, copy, delete
+
+.. note::
+    Tests do not check all attributes of returned objects as deserialising
+    objects is not directly a SALClient responsibility
+"""
+
+from datetime import datetime
+from unittest.mock import patch
+
+import pytest
+
+from sal.core.serialise import serialise
+from sal.core.object import BranchReport, Branch
+
+
+@pytest.fixture
+def timestamp():
+
+    return datetime(2000, 1, 1, 12, 00, 00)
+
+
+@pytest.fixture
+def branch_report_response(server_response, timestamp):
+
+    """
+    JSON representing a BranchReport
+    """
+
+    br = BranchReport('A BranchReport', ['b1', 'b2', 'b3'], [], timestamp)
+    server_response.json.return_value = serialise(br)
+    return server_response
+
+
+@pytest.fixture
+def branch():
+
+    return Branch('A Branch')
+
+@pytest.fixture
+def branch_response(branch, server_response):
+
+    """
+    JSON representing a Branch
+    """
+
+    server_response.json.return_value = serialise(branch)
+    return server_response
+
+
+@pytest.fixture
+def auth_header(token):
+
+    return {'Authorization': 'Bearer 78945JHKJFSJDFKH7897wej8UIOJKhuwiofdSDHk'}
+
+
+@pytest.mark.parametrize('path, revision, verify_https_cert',
+                         [('/node', 0, False),
+                          ('/node:head', 0, True),
+                          ('/node:10', 10, False)])
+def test_list(host, patched_client, timestamp, branch_report_response,
+              path, revision, verify_https_cert):
+
+    """
+    Parametrization includes the three different ways of passing a revision
+
+    GIVEN
+        A server with data nodes
+    WHEN
+        A client performs a list operation on a path
+    THEN
+        The client returns a BranchReport
+    """
+
+    patched_client.verify_https_cert = verify_https_cert
+
+    with patch('sal.client.main.requests.get',
+               return_value=branch_report_response) as brr:
+        br = patched_client.list(path)
+
+        brr.assert_called_with('{0}/data/node?revision={1}'.format(host,
+                                                                   revision),
+                               verify=verify_https_cert)
+        assert isinstance(br, BranchReport)
+        assert br.timestamp == timestamp
+
+
+@pytest.mark.parametrize('path, revision, verify_https_cert, summary',
+                         [('/node', 0, False, True),
+                          ('/node:head', 0, True, False),
+                          ('/node:10', 10, False, False)])
+def test_get(host, patched_client, timestamp, branch_response, path, revision,
+             verify_https_cert, summary):
+
+    """
+    GIVEN
+        A server with data nodes
+    WHEN
+        A client performs a get operation
+    THEN
+        The client returns either a Branch or a DataObject
```

Comment:
**question:** The actual test client only returns a Branch, and only a Branch is ever checked for. Is this specification overbroad, or is the test not fully implemented?

Reply:
Good point, I must have forgotten to add the parametrization for testing a `DataObject`.  I've done so using a `Scalar` and I've also added the same type of parametrization for `list` and `put`, as these both really should apply to both the contents of branches and leaves, and there are minor differences in how the `SALClient` handles this.